### ISSUE TITLE
item_features are not actually used in the tutorial, and other questions. 

### ISSUE THREAD

**Issue by soliverc** [etlundquist/rankfm#29]
Just a few things I'm not clear on going through the medium article (https://towardsdatascience.com/factorization-machines-for-item-recommendation-with-implicit-feedback-data-5655a7c749db) and reading the notebook here on github. 

**1.** 

The medium talks about using item / user features as part of a recommender system. The accompanying notebook creates item features based on the aisle number:

`item_features` 
`item_features_train`
`item_features_valid`

and never uses them. Unless I missed something? I check the docs also. Should the notebook be using them somewhere?

Should the product_id be in the index of `item_features_train`? Because in the notebook it's just a regular column: 

![image](https://user-images.githubusercontent.com/47222261/97977952-d8060200-1dc4-11eb-8a8f-03deff70be51.png)

Anyway, it looks like `item_features_train` goesinto `model.fit()` 

```python

model.fit(interactions_train, sample_weight=sample_weight_train, epochs=30, verbose=True, item_features=item_features_train)

```

I've compared the evaluation using `item_features_train` to the original model in the notebook which doesn't use  `item_features_train`. The scores are slightly lower when I use `item_features` but not by much. Which makes me think I'm either doing something wrong (should `product_id` be in the index of `item_features_train`?), or these features are just not great and don't add anything other than noise. 

![image](https://user-images.githubusercontent.com/47222261/97896148-988bd700-1d2c-11eb-8a34-c2a6ad2f64fd.png)

**2.**

The dataset also has `department id` as a potential feature. How would we use both aisle and department id in the same model? The `item_features` argument in `model.fit()` takes a single dummified dataframe. Do we use `fit_partial` to update the model with extra dataframes of user features or item features? Let's say I have the age, city, gender, and monthly income of my users. This would be four dataframes. How would I use them all?

**3.** 

What is item_features_valid used for? This is created in the notebook but never used. 

**4.** 

There is another section where scores are generated:

```python

scores = model.predict(interactions_valid, cold_start='nan')

print(scores)

array([-1.1237175 ,  0.00314923,  1.5434608 , ...,  2.029984  ,
        1.8916801 ,  2.7111115 ], dtype=float32)

```
What are these used for? They are not used in the notebook after generating them.

**Comment by etlundquist** [comment]
Hi @soliverc - first, i'm sorry it's taken me so long to respond to you - I got really really busy with my regular job this winter and didn't have any time to work on this package or even really keep track of what was going on with user issues.

(1) there may be differences between the GitHub notebooks and the Medium article - those were never intended to be 1:1 in sync. However, the call that you posted above looks correct:

```
model.fit(interactions_train, sample_weight=sample_weight_train, epochs=30, verbose=True, item_features=item_features_train)
```

I tried to write this in the docstring of the `fit()` method ([docs (https://rankfm.readthedocs.io/en/latest/rankfm.html#rankfm.rankfm.RankFM.fit)), but the index doesn't matter - the package is assuming that the first column of the side-feature dataframes is the `user_id` or `item_id`, and that the set of IDs in the side features is the same as the set of IDs in the interactions data. I'm pretty sure the math is right for using these side features, but maybe it's the case that they don't add a whole lot above and beyond the user-item interactions for the instacart data. 

(2) you need to put user or item features in a single dataframe at the `user_id` or `item_id` level. Let's take `department_id` as an example. you could include that in the `item_features` dataframe by simply joining each `product_id` to its corresponding `department_id`. A lot of the values are going to be repeated, but you have to supply your side features in this de-normalized format. Same goes for user features. You still need a single dataframe at the `user_id` level, but you could easily tack on features like `age`, `city`, `gender`, ... by just joining them to the features frame by `user_id`. Again, the docstring of the `fit()` method tries to provide a bit more info on the expected structure. Instead of using 4 separate dataframes with different types of features at different normalization levels, you just have to merge them into one DF at the user_id level, and/or one DF at the item_id level. Should be possible for any user/item features you have. 

(3) nothing - you don't really need `item_features_valid`. The reason why is that during calls to `model.predict()` it assumes you are calling it for users/items that were present during fitting. If that's the case then you already have the features for those users and/or items stored from the call to fit(), so there's no need to re-supply. You can ignore this.

(4) just to demonstrate what the output of `predict()` looks like. Sometimes you need to generate topN recommendations for each user, but other times (this is very common in my experience) you need to score a specific set of user/item pairs. Maybe it's the case that you've pre-filtered users/items or have additional business logic you need to enforce, so you can't just take the topN items for each user. If you only ever need the `recommend()` method that's fine too. 

Hope this helps! thanks for using the package, and hopefully late is better than never to answer your questions! Take care.
